apps/bixarena/auth-service/src/main/java/org/sagebionetworks/bixarena/auth/service/service/UserService.java

```diff
@@ -0,0 +1,169 @@
+   *
+   * @param provider The OAuth2 provider (e.g., synapse)
+   * @param externalId The user's ID from the external provider
+   * @param username The preferred username
+   * @param email The user's email address
+   * @param firstName The user's first name (given name)
+   * @param lastName The user's last name (family name)
+   * @return The UserEntity representing the logged-in user
+   */
+  @Transactional
+  public UserEntity handleUserLogin(
+    Provider provider,
+    String externalId,
+    String username,
+    String email,
+    String firstName,
+    String lastName
+  ) {
+    // Try to find existing external account first
+    Optional<ExternalAccountEntity> existingExtAccount =
+      externalAccountRepository.findByProviderAndExternalId(provider, externalId);
+
+    UserEntity user;
+
+    if (existingExtAccount.isPresent()) {
+      // User has logged in before with this provider
+      user = existingExtAccount.get().getUser();
+      log.info(
+        "Existing user login: userId={} username={} provider={} externalId={}",
+        user.getId(),
+        user.getUsername(),
+        provider,
+        externalId
+      );
+
+      // Update user information in case it changed at the provider
+      boolean updated = false;
+      if (email != null && !email.equals(user.getEmail())) {
+        user.setEmail(email);
+        updated = true;
+      }
+      if (firstName != null && !firstName.equals(user.getFirstName())) {
+        user.setFirstName(firstName);
+        updated = true;
+      }
+      if (lastName != null && !lastName.equals(user.getLastName())) {
+        user.setLastName(lastName);
+        updated = true;
+      }
+
+      // Update last login timestamp
+      user.setLastLoginAt(OffsetDateTime.now());
+      user = userRepository.save(user);
+
+      if (updated) {
+        log.info("Updated user information for userId={}", user.getId());
+      }
+
+      // Update external account information
+      ExternalAccountEntity extAccount = existingExtAccount.get();
+      boolean extUpdated = false;
+      if (username != null && !username.equals(extAccount.getExternalUsername())) {
+        extAccount.setExternalUsername(username);
+        extUpdated = true;
+      }
+      if (email != null && !email.equals(extAccount.getExternalEmail())) {
+        extAccount.setExternalEmail(email);
+        extUpdated = true;
+      }
+      if (extUpdated) {
+        externalAccountRepository.save(extAccount);
+      }
+    } else {
+      // New user - check if username already exists (shouldn't happen, but defensive)
+      Optional<UserEntity> existingUser = userRepository.findByUsername(username);
+
+      if (existingUser.isPresent()) {
+        user = existingUser.get();
+        log.info(
+          "Linking existing user to new provider: userId={} username={} provider={} externalId={}",
+          user.getId(),
+          user.getUsername(),
+          provider,
+          externalId
+        );
```

Comment:
just curious: are we checking the user's username from different providers to determine if they are the same user?